src/persist-client/src/lib.rs

```diff
@@ -136,14 +159,24 @@ impl Client {
         D: Semigroup + Codec64,
     {
         trace!("Client::open timeout={:?} id={:?}", timeout, id);
-        let mut shards = self.shards.lock().await;
-        let state = shards
-            .entry(id)
-            .or_insert_with(|| Shard::<K, V, T, D>::new(id.clone()).into_inner());
-        let state = Shard::decode(Arc::clone(state))
-            .await
-            .expect("internal error: newly created shard's codecs don't match");
-        Ok(state.recover_capabilities().await)
+        let deadline = Instant::now() + timeout;
+        let mut machine = Machine::new(id, Arc::clone(&self.consensus));
+        let (writer_id, reader_id) = (WriterId::new(), ReaderId::new());
+        let (write_cap, read_cap) = machine.register(deadline, &writer_id, &reader_id).await?;
```

Comment:
on reflection, I think the right place to allocate the ids is inside Machine::register (with State::register still taking them as deterministic inputs). put another way, pushing them down a level